File: cpp/bob/bob.cpp
#include "bob.h"
#include <regex>

using std::regex;

const char* bob::hey(const char* utterance) {
  if (regex_search(utterance, regex("[A-Z]")) &&
      !regex_search(utterance, regex("[a-z]"))) {
    return "Whoa, chill out!";
  } else if (regex_match(utterance, regex(".*\\?\\s*"))) {
    return "Sure.";
  } else if (regex_match(utterance, regex("\\s*"))) {
    return "Fine. Be that way!";
  } else {
    return "Whatever.";
  }
};

File: haskell/bob/src/Bob.hs
module Bob (responseFor) where
import Text.Regex.PCRE ((=~))

isShouting :: String -> Bool
isShouting utterance = (utterance =~ "[A-Z]") &&
                       (not (utterance =~ "[a-z]"))

isAsking :: String -> Bool
isAsking utterance = utterance =~ "\\?\\s*\\Z"

isSilence :: String -> Bool
isSilence utterance = utterance =~ "\\A\\s*\\Z"

responseFor :: String -> String
responseFor utterance
  | isShouting utterance = "Whoa, chill out!"
  | isAsking utterance   = "Sure."
  | isSilence utterance  = "Fine. Be that way!"
  | otherwise            = "Whatever."
